sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1188,8 +1188,29 @@ class Dataset[T] private[sql](
     case "*" =>
       Column(ResolvedStar(queryExecution.analyzed.output))
     case _ =>
-      val expr = resolve(colName)
-      Column(expr)
+      if (sqlContext.conf.supportQuotedRegexColumnName) {
+        colRegex(colName)
+      } else {
+        val expr = resolve(colName)
+        Column(expr)
+      }
+  }
+
+  /**
+   * Selects column based on the column name specified as a regex and return it as [[Column]].
+   * @group untypedrel
+   * @since 2.3.0
+   */
+  def colRegex(colName: String): Column = {
```

Comment:
`col` returns a column resolved on the current `Dataset`. `colRegex` now can return an unresolved one. Seems ok, but any possibility to go wrong with it?

For example, we can do:

    val colRegex1 = df1.colRegex("`...`")  // colRegex1 is unresolved.
    df2.select(colRegex1)

But you can't do the same thing with `col`.